docs/language/learn-ql/cpp/basic-query-cpp.rst

```diff
@@ -0,0 +1,145 @@
+| ``select ifstmt, "This 'if' statement is redundant."``        | Defines what to report for each match.                                                                            | Reports the resulting ``if`` statement with a string that explains the problem.                                        |
+|                                                               |                                                                                                                   |                                                                                                                        |
+|                                                               | ``select`` statements for queries that are used to find instances of poor coding practice are always in the form: |                                                                                                                        |
+|                                                               | ``select <program element>, "<alert message>"``                                                                   |                                                                                                                        |
++---------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------+
+
+Extend the query
+----------------
+
+Query writing is an inherently iterative process. You write a simple query and then, when you run it, you discover examples that you had not previously considered, or opportunities for improvement.
+
+Remove false positive results
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Browsing the results of our basic query shows that it could be improved. Among the results you are likely to find examples of ``if`` statements with an ``else`` branch, where an empty ``then`` branch does serve a purpose. For example:
+
+.. code-block:: cpp
+
+   if (...) {
+     ...
+   } else if (!strcmp(option, "-verbose") {
+     // nothing to do - handled earlier
+   } else {
+     error("unrecognized option");
+   }
+
+In this case, identifying the ``if`` statement with the empty ``then`` branch as redundant is a false positive. One solution to this is to modify the query to ignore empty ``then`` branches if the ``if`` statement has an ``else`` branch.
+
+To exclude ``if`` statements that have an ``else`` branch:
+
+#. Extend the ``where`` clause to include the following extra condition:
+
+   .. code-block:: ql
+
+      and not ifstmt.hasElse()
+
+   The ``where`` clause is now:
+
+   .. code-block:: ql
+
+      where ifstmt.getThen() = block and
+        block.getNumStmt() = 0 and
+        not ifstmt.hasElse()
```

Comment:
The code and query linked in the previous section and the query linked below have the `and`s at the beginning of the line.
Maybe it would be good to place them at the beginning here as well for consistency.

Reply:
Thanks for pointing that out 😃 I've actually moved all the `and`s to the end of the line (in accordance with [this section of the style guide](https://github.com/github/codeql/blob/master/docs/ql-style-guide.md#formulas)).